[Connectivity] add a method to request location on iOS (for iOS 13)

cyanglaz opened the pull request:
## Description

Starting from iOS 13. Certain apps need to have user authorized location service to get WIFI information using `CNCopyCurrentNetworkInfo`. Detailed in https://github.com/flutter/flutter/issues/37804

We add a helper method for the plugin users to request or get location service authorization.
Also updated the REAMDE to explain the iOS 13 changes.

## Related Issues

https://github.com/flutter/flutter/issues/37804

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

collinjackson commented on packages/connectivity/README.md:
```diff
@@ -60,17 +60,39 @@ var wifiIP = await (Connectivity().getWifiIP());network
+### iOS 12
 
-#### iOS 13
+To use `.getWifiBSSID()` and `.getWifiName()` on iOS >= 12, the `Access WiFi information capability` in XCode must be enabled. Otherwise, both method will return null.
```
```suggestion
To use `.getWifiBSSID()` and `.getWifiName()` on iOS >= 12, the `Access WiFi information capability` in XCode must be enabled. Otherwise, both methods will return null.
```

collinjackson commented on packages/connectivity/README.md:
```diff
@@ -60,17 +60,39 @@ var wifiIP = await (Connectivity().getWifiIP());network
+
+ * BSSID: "00:00:00:00:00:00"
+
+An App linked against iOS 13 or later receives null.
```
```suggestion
An app linked against iOS 13 or later receives `null`.
```

collinjackson commented on packages/connectivity/README.md:
```diff
@@ -60,17 +60,39 @@ var wifiIP = await (Connectivity().getWifiIP());network
-> SSID: "Wi-Fi" or "WLAN" ("WLAN" will be returned for the China SKU)  
-> BSSID: "00:00:00:00:00:00"
+As of iOS 13, Apple announced that these APIs will no longer return valid information.
+An App linked against iOS 12 or earlier receives pseudo-values such as:
```
```suggestion
An app linked against iOS 12 or earlier receives pseudo-values such as:
```

collinjackson commented on packages/connectivity/example/ios/Runner/Info.plist:
```diff
@@ -22,6 +22,10 @@
+	<key>NSLocationAlwaysAndWhenInUseUsageDescription</key>
+	<string>Location service</string>
+	<key>NSLocationWhenInUseUsageDescription</key>
+	<string>locaiton service in use</string>
```
Typo, also I think you should put an English sentence here

collinjackson commented on packages/connectivity/example/ios/Runner/Info.plist:
```diff
@@ -22,6 +22,10 @@
 	<key>LSRequiresIPhoneOS</key>
 	<true/>
+	<key>NSLocationAlwaysAndWhenInUseUsageDescription</key>
+	<string>Location service</string>
```
I think you should put an appropriate English sentence here since it's an example app.

collinjackson commented on packages/connectivity/lib/connectivity.dart:
```diff
@@ -93,6 +94,66 @@ class Connectivity {
+  /// It will show a platform standard window of requesting a location service.
+  ///
+  /// If the user declined the location service, it will never show the window to request the authorization again.
+  /// The user has to go to the setting app in the phone to enable authorization.
```
```suggestion
  /// The user has to go to the settings app in the phone to enable authorization.
```

collinjackson commented on packages/connectivity/lib/connectivity.dart:
```diff
@@ -93,6 +94,66 @@ class Connectivity {
+  /// If the user declined the location service, it will never show the window to request the authorization again.
+  /// The user has to go to the setting app in the phone to enable authorization.
+  ///
+  /// This method is a helper to get the location authorization that is necessary for certain functionalities in this plugin.
```
```suggestion
  /// This method is a helper to get the location authorization that is necessary for certain functionality of this plugin.
```

collinjackson commented on packages/connectivity/lib/connectivity.dart:
```diff
@@ -93,6 +94,66 @@ class Connectivity {
+  /// This request.
+  ///
+  /// if the location information needs to be accessible all the time, set `requestAlwaysLocationUsage` to true. Note that the status
+  /// returned might not be [LocationAuthorizationStatus.authorizedAlways] even you requested it. User might have already chosen a location authorization
```
```suggestion
  /// returned might not be [LocationAuthorizationStatus.authorizedAlways] even if you requested it. The user might have already chosen a location authorization
```

collinjackson commented on packages/connectivity/lib/connectivity.dart:
```diff
@@ -93,6 +94,66 @@ class Connectivity {
+  /// Request to authorize the location service. Only on iOS.
+  ///
+  /// Returns a [LocationAuthorizationStatus] if the location service has already been authorized or user authorized the location on
+  /// This request.
```
```suggestion
  /// this request.
```

collinjackson commented on packages/connectivity/README.md:
```diff
@@ -60,17 +60,39 @@ var wifiIP = await (Connectivity().getWifiIP());network
+
+An App linked against iOS 13 or later receives null.
+
+The [CNCopyCurrentNetworkInfo] will work for Apps that:
```
```suggestion
The `CNCopyCurrentNetworkInfo` will work for Apps that:
```

collinjackson commented on packages/connectivity/README.md:
```diff
@@ -60,17 +60,39 @@ var wifiIP = await (Connectivity().getWifiIP());network
+
+### iOS 13
 
 The methods `.getWifiBSSID()` and `.getWifiName()` utilize the [CNCopyCurrentNetworkInfo](https://developer.apple.com/documentation/systemconfiguration/1614126-cncopycurrentnetworkinfo) function on iOS.
```
```suggestion
The methods `.getWifiBSSID()` and `.getWifiName()` utilize the [`CNCopyCurrentNetworkInfo`](https://developer.apple.com/documentation/systemconfiguration/1614126-cncopycurrentnetworkinfo) function on iOS.
```

collinjackson commented on packages/connectivity/README.md:
```diff
@@ -60,17 +60,39 @@ var wifiIP = await (Connectivity().getWifiIP());network
+To request location authorization, make sure to add the following keys to your _Info.plist_ file, located in `<project root>/ios/Runner/Info.plist`:
 
-You can follow issue [#37804](https://github.com/flutter/flutter/issues/37804) for the changes required to return valid SSID and BSSID values with iOS 13.
+* `NSLocationAlwaysAndWhenInUseUsageDescription` - describe why the app needs access to the user’s location information all the time (foreground and background). _Privacy - Location Always and When In Use Usage Description_ in the visual editor.
```
```suggestion
* `NSLocationAlwaysAndWhenInUseUsageDescription` - describe why the app needs access to the user’s location information all the time (foreground and background). This is called _Privacy - Location Always Usage Description_ in the visual editor.
```

collinjackson replied:
I might be wrong here but I think it should just be _Privacy - Location Always Usage Description_

collinjackson commented on packages/connectivity/ios/Classes/ConnectivityPlugin.m:
```diff
@@ -97,6 +103,41 @@ - (NSString*)statusFromReachability:(Reachability*)reachability {
+  CLAuthorizationStatus status = CLLocationManager.authorizationStatus;
+  switch (status) {
+    case kCLAuthorizationStatusNotDetermined: {
+      self.result = result;
```
I think there's a race condition here where two calls to `requestLocationServiceAuthorizationIfUndetermined` in quick succession could lead to result being clobbered before being invoked. Consider returning an error if self.result != null

collinjackson commented on packages/connectivity/ios/Classes/ConnectivityPlugin.m:
```diff
@@ -140,4 +184,18 @@ - (FlutterError*)onCancelWithArguments:(id)arguments {
+}
+
+- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
+  [self requestLocationServiceAuthorizationIfUndetermined:self.result always:nil];
```
Should probably set `self.result` to `nil` here so we don't accidentally notify it more than once.

cyanglaz commented:
>However, it feels like we're conflating requesting permission and checking what the current status is. I'm wondering if there should be separate APIs for each. What do you think?

Sure. My initial implementation did actually have these separated. I can add it back.

cyanglaz commented:
@collinjackson Thanks for the review. I updated the code with your review suggests; also added a driver test to test the `getLocationServiceAuthorization`. The other method `requestLocationServiceAuthorization` is probably not testable as it for now since it will show a pop up and I'm not sure if there is a way to handle system pop ups from flutter driver.
Im going to merge this on green if you don't have any objections.
